File: c/clang/_demo/castdump/astdump.go
package main

import (
	"fmt"
	"os"

	"github.com/goplus/llgo/c"
	"github.com/goplus/llgo/c/clang"
)

func visit(cursor, parent clang.Cursor, clientData c.Pointer) clang.ChildVisitResult {
	depth := *(*c.Uint)(clientData)
	printAST(cursor, depth+1)
	return clang.ChildVisit_Continue
}

func printAST(cursor clang.Cursor, depth c.Uint) {
	cursorKind := cursor.Kind.String()

	cursorSpelling := cursor.String()

	for i := c.Uint(0); i < depth; i++ {
		c.Fputs(c.Str("  "), c.Stdout)
	}

	c.Printf(c.Str("%s: %s\n"), cursorKind.CStr(), cursorSpelling.CStr())

	cursorKind.Dispose()
	cursorSpelling.Dispose()

	clang.VisitChildren(cursor, visit, c.Pointer(&depth))
}

func main() {
	if c.Argc != 2 {
		fmt.Fprintln(os.Stderr, "Usage: castdump <headerFile>")
		return
	}
	sourceFile := *c.Advance(c.Argv, 1)

	index := clang.CreateIndex(0, 0)

	unit := index.ParseTranslationUnit(
		sourceFile,
		nil, 0,
		nil, 0,
		clang.TranslationUnit_None,
	)

	if unit == nil {
		println("Unable to parse translation unit. Quitting.")
		c.Exit(1)
	}

	cursor := unit.Cursor()

	printAST(cursor, 0)

	unit.Dispose()
	index.Dispose()
}

File: compiler/chore/_xtool/astdump/astdump.cpp
#include <clang-c/Index.h>
#include <iostream>
#include <string>

CXChildVisitResult visit(CXCursor c, CXCursor parent, CXClientData client_data);

void printAST(CXCursor cursor, unsigned int depth = 0) {
    CXString cursorKind = clang_getCursorKindSpelling(clang_getCursorKind(cursor));
    CXString cursorSpelling = clang_getCursorSpelling(cursor);

    for (unsigned int i = 0; i < depth; ++i) {
        std::cout << "  ";
    }

    std::cout << clang_getCString(cursorKind) << ": " 
              << clang_getCString(cursorSpelling) << std::endl;

    clang_disposeString(cursorKind);
    clang_disposeString(cursorSpelling);

    CXCursor child;
    clang_visitChildren(
        cursor,
        visit,
        &depth
    );
}

CXChildVisitResult visit(CXCursor c, CXCursor parent, CXClientData client_data) {
    unsigned int* depth = (unsigned int*)client_data;
    printAST(c, *depth + 1);
    return CXChildVisit_Continue;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <header_file>" << std::endl;
        return 1;
    }

    CXIndex index = clang_createIndex(0, 0);
    CXTranslationUnit unit = clang_parseTranslationUnit(
        index,
        argv[1],
        nullptr, 0,
        nullptr, 0,
        CXTranslationUnit_None
    );

    if (unit == nullptr) {
        std::cerr << "Unable to parse translation unit. Quitting." << std::endl;
        return 1;
    }

    CXCursor cursor = clang_getTranslationUnitCursor(unit);
    std::cout << "AST for " << argv[1] << ":" << std::endl;
    printAST(cursor);

    clang_disposeTranslationUnit(unit);
    clang_disposeIndex(index);

    return 0;
}